// mirada/dist/src/format/format.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var misc_utils_of_mine_generic_1 = require("misc-utils-of-mine-generic");
var file_1 = require("../file");
/**
 * Nor or opencv.js or this library implement any image format so users are
 * responsible of providing a FormatProxy using some library.
 *
 */
function installFormatProxy(proxy) {
    return __awaiter(this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            proxies.push(proxy);
            return [2 /*return*/];
        });
    });
}
exports.installFormatProxy = installFormatProxy;
function unInstallFormatProxies() {
    return __awaiter(this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            proxies.length = 0;
            return [2 /*return*/];
        });
    });
}
exports.unInstallFormatProxies = unInstallFormatProxies;
var proxies = [];
var codecs = [];
var _proxyLoaded = false;
function createCodec(proxy) {
    return __awaiter(this, void 0, void 0, function () {
        var _a;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    if (!(typeof proxy === 'function')) return [3 /*break*/, 2];
                    return [4 /*yield*/, proxy()];
                case 1:
                    _a = _b.sent();
                    return [3 /*break*/, 4];
                case 2: return [4 /*yield*/, proxy.create()];
                case 3:
                    _a = _b.sent();
                    _b.label = 4;
                case 4: return [2 /*return*/, _a];
            }
        });
    });
}
/**
 * @internal
 */
function loadFormatProxies() {
    return __awaiter(this, void 0, void 0, function () {
        var _a;
        var _this = this;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    if (!!_proxyLoaded) return [3 /*break*/, 3];
                    _proxyLoaded = true;
                    _a = proxies.length;
                    if (!_a) return [3 /*break*/, 2];
                    return [4 /*yield*/, misc_utils_of_mine_generic_1.serial(proxies.map(function (proxy) { return function () { return __awaiter(_this, void 0, void 0, function () {
                            var p;
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0: return [4 /*yield*/, createCodec(proxy)];
                                    case 1:
                                        p = _a.sent();
                                        codecs.push(p);
                                        return [2 /*return*/];
                                }
                            });
                        }); }; }))];
                case 1:
                    _a = (_b.sent());
                    _b.label = 2;
                case 2:
                    _a;
                    _b.label = 3;
                case 3: return [2 /*return*/];
            }
        });
    });
}
exports.loadFormatProxies = loadFormatProxies;
function unloadFormatProxies() {
    codecs.length = 0;
}
exports.unloadFormatProxies = unloadFormatProxies;
function getDefaultCodec() {
    var c = codecs.length ? codecs[0] : undefined;
    if (!c) {
        throw new Error('No codec found. you need to provide a proxy and wait for loadFormatProxies()');
    }
    return c;
}
exports.getDefaultCodec = getDefaultCodec;
function decodeOrThrow(buffer, format) {
    return __awaiter(this, void 0, void 0, function () {
        var r;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, getDefaultCodec().decode(buffer, format)];
                case 1:
                    r = _a.sent();
                    misc_utils_of_mine_generic_1.checkThrow(r, "Fail to decode buffer. " + (format ? "requested format: " + format : '') + ". Detected format: " + (file_1.File.getBufferFileType(buffer) && file_1.File.getBufferFileType(buffer).mime || 'unknown'));
                    return [2 /*return*/, r];
            }
        });
    });
}
exports.decodeOrThrow = decodeOrThrow;
function encodeOrThrow(data, format, quality) {
    return __awaiter(this, void 0, void 0, function () {
        var r;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, getDefaultCodec().encode(data, format, quality)];
                case 1:
                    r = _a.sent();
                    misc_utils_of_mine_generic_1.checkThrow(r, 'Fail to encode to requested format ' + format + '. Given: ' + format);
                    return [2 /*return*/, r];
            }
        });
    });
}
exports.encodeOrThrow = encodeOrThrow;
