// mirada/dist/src/browser/index.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
var cameraHelper = __importStar(require("./cameraHelper"));
var canvasRender = __importStar(require("./canvasRender"));
var imageCreation = __importStar(require("./imageCreation"));
var VideoReader = __importStar(require("./videoReader"));
var cameraHelper_1 = require("./cameraHelper");
exports.CameraHelper = cameraHelper_1.CameraHelper;
var canvasRender_1 = require("./canvasRender");
exports.renderArrayBufferInCanvas = canvasRender_1.renderArrayBufferInCanvas;
exports.renderInCanvas = canvasRender_1.renderInCanvas;
var imageCreation_1 = require("./imageCreation");
exports.getHtmlImageData = imageCreation_1.asHtmlImageData;
exports.fetchImageData = imageCreation_1.fetchImageData;
exports.fromInputFileElement = imageCreation_1.fromInputFileElement;
var videoReader_1 = require("./videoReader");
exports.VideoReader = videoReader_1.VideoReader;
exports.browser = __assign(__assign(__assign(__assign({}, canvasRender), cameraHelper), imageCreation), VideoReader);
